torchacc/dist/fsdp.py

```diff
@@ -202,3 +219,178 @@ def clip_grad_norm_(self, max_grad_norm):
+            ``state_dict_type``.
+        """
+        # we only support FULL_STATE_DICT and flatten parameters now
+        if state_dict_type != StateDictType.FULL_STATE_DICT:
+            raise NotImplementedError(
+                "we only support 'FULL_SATE_DICT' StateDictType now")
+        if not self.model.flatten_parameters:
+            raise NotImplementedError(
+                "we only support flatten_parameters=True now")
+
+        shard_meta_data = self.model.get_shard_metadata()
+        sharded_optim_state = optim.state_dict()['state']
+        optim_state_param_groups = optim.state_dict()['param_groups']
+        # unflattened and consolidated state_dict
+        consolidate_optim_state_dict: Dict[str, Any] = {
+            'state': {},
+            'param_groups': {}
+        }
+
+        # (rank0_only and self.model.rank == 0) or (not rank0_only)
+        if not rank0_only or self.model.rank == 0:
+            consolidate_optim_state_dict[
+                'param_groups'] = optim_state_param_groups
+            consolidate_optim_state_dict['param_groups'][0]['params'].clear()
+
+        for layer_state, (layer_name, layer_params) in zip(
+                sharded_optim_state.values(),
+                shard_meta_data['flatten_info'].items()):
+            param_names, param_shapes, param_numels = layer_params
+            # get full_param_names of each layer in optim_state_dict
+            full_names = optim_utils._get_layer_full_names(
+                layer_name, param_names)
+            unflat_state_dict = {fn: {} for fn in full_names}
+
+            if not rank0_only or self.model.rank == 0:
+                consolidate_optim_state_dict['param_groups'][0][
+                    'params'].append(full_names)
+
+            for state_name, state_params in layer_state.items():
+                tensor_buffer = optim_utils._all_gather_state(
+                    state_params, self.model)
+
+                if not rank0_only or self.model.rank == 0:
+                    _, full_params = optim_utils._unflatten_optim_params(
+                        tensor_buffer, layer_name, param_names, param_shapes,
+                        param_numels)
+
+                    for fn, fp in zip(full_names, full_params):
+                        if cpu_offload:
+                            xm.mark_step()
+                            unflat_state_dict[fn][state_name] = fp.cpu()
+                        else:
+                            unflat_state_dict[fn][state_name] = fp
+                xm.mark_step()
+        consolidate_optim_state_dict['state'] = unflat_state_dict
+
+        return consolidate_optim_state_dict
+
+    def load_optim_state_dict(
+            self,
+            optim_state_dict: Dict[str, Any],
+            optim: torch.optim.Optimizer,
+            state_dict_type: StateDictType = StateDictType.FULL_STATE_DICT,
+            rank0_only: bool = True) -> Dict[str, Any]:
```

Comment:
also, can we use the same api as https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.optim_state_dict_to_load ?

Reply:
can we directly check the type of optim?

Reply:
we can use the api to load full_optimizer state_dict. I think  one possible way to check the type of optim is to compare the numel of params of state_dict of one layer with that stored in sharded_meta_data.